From a3894b84e49383c00f556c11d92061037c0674f8 Mon Sep 17 00:00:00 2001
From: Amir Vadai <avadai@redhat.com>
Date: Thu, 11 Sep 2014 12:01:44 -0400
Subject: [ethernet] mlx4: cq->irq_desc wasn't set in legacy EQ's

Message-id: <f083418a5a439944d2f9f3c424b627f93f9f9627.1410425016.git.avadai@redhat.com>
Patchwork-id: 93721
O-Subject: [Patch RHEL7.1 V1 168/197] net/mlx4_en: cq->irq_desc wasn't set in legacy EQ's
Bugzilla: 1107617 1107618 1107619 1107620
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>

From: Amir Vadai <amirv@mellanox.com>

Fix a regression introduced by commit 35f6f45 ("net/mlx4_en: Don't use
irq_affinity_notifier to track changes in IRQ affinity map").
When core is started in legacy EQ's (number of IRQ's < rx rings), cq->irq_desc
was NULL.  This caused a kernel crash under heavy traffic - when having more
than rx NAPI budget completions.
Fixed to have it set for both EQ modes.

Signed-off-by: Amir Vadai <amirv@mellanox.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit 858e6c321065344339906672bccd0eafe9622258)
Signed-off-by: Amir Vadai <avadai@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/mellanox/mlx4/en_cq.c b/drivers/net/ethernet/mellanox/mlx4/en_cq.c
index 6b2e160..5a19a55 100644
--- a/drivers/net/ethernet/mellanox/mlx4/en_cq.c
+++ b/drivers/net/ethernet/mellanox/mlx4/en_cq.c
@@ -130,16 +130,17 @@ int mlx4_en_activate_cq(struct mlx4_en_priv *priv, struct mlx4_en_cq *cq,
         name);
     }
 
-#ifdef CONFIG_GENERIC_HARDIRQS
-    cq->irq_desc =
-     irq_to_desc(mlx4_eq_get_irq(mdev->dev,
-            cq->vector));
-#endif
    }
   } else {
    cq->vector = (cq->ring + 1 + priv->port) %
     mdev->dev->caps.num_comp_vectors;
   }
+
+#ifdef CONFIG_GENERIC_HARDIRQS
+  cq->irq_desc =
+   irq_to_desc(mlx4_eq_get_irq(mdev->dev,
+          cq->vector));
+#endif
  } else {
   /* For TX we use the same irq per
   ring we assigned for the RX    */
-- 
1.7.1